### DISCUSSION TITLE
How to complete AzerothCore 3.0

### DISCUSSION THREAD

**Discussion by Yehonal** [azerothcore/azerothcore-wotlk#3900] [General]
How to complete AzerothCore 3.0

As many of you noticed, most of the mangos-based servers are using the strategy of defining themself "unfinished-software".
The main reason is that such software are constantly a work-in-progress to fix bugs and implement new features at the same time that eventually generate more bugs and everything on the same branch.

# The road to a real stable version

## Semver:

**DONE**: To do a step forward AzerothCore adopted the **Semver** project versioning standard from the beginning: https://www.azerothcore.org/wiki/Project-Versioning

**TODO**: 
- deciding a versioning strategy for the database squash
- using a proper way to versioning patches (fix), minor (non breaking-features), major (breaking features). At the moment we're only tracking the "major" part.
- creating "How to upgrade" articles in our wiki. The one for the dev and the one for the users
- continuing to fix security and stability fixes to older supported branches (since the 4.0)
- Using a proper way to check compatibility between modules and AzerothCore during the installation/compilation process.

**Important NOTE:** The Semver will help us to keep the compatibility with the [ecosystem ](https://github.com/azerothcore/forum/issues/99) that we're creating. So it's very important to start using it in the correct way.

## Documentation

**DONE**: Our wiki is the single-source-of-truth of all our server documentation: https://www.azerothcore.org/wiki/home . It's based on git and full of articles, however, it's just the beginning!

**TODO**:
- completing the wiki regarding the database, the modular system etc. The project is very huge and it's difficult to document all the features that come with the AzerothCore. It's really important that everyone can start on the basics and then extends their knowledge about the AC reading the code. 
- Setting auto-generated code documentation and/or a better online code browser
- Document the functional requirements of the "product" itself. It means that we NEED to grab all information about spell formulas, NPCs etc. from the web... putting them on this website: https://wowgaming.github.io/ before they disappear

## Testing

Maybe the most important part! if we want to complete the AzerothCore-Wotlk product and be sure that the quality of the project is preserved during the time, the testing layer is the ONLY solution (I'll explain it to you why at the end of this post).

 **TODO**:
- Create the first Unit test for one of our method
- Create the first Integration test for one of our method
- Generate the coverage files and publish on gh-pages
- Complete and use our [web-based client](https://github.com/azerothcore/acore-client) to create a workflow of e2e tests that will send opcodes and will check the response. 
- integrate everything inside our CI (github action)
- AzerothCore integrated module to collect anonymous information during the runtime, so we can keep automatically keep track of certain issues.

Personal considerations about testing and its importance:

AzerothCore is not a simple application that can be unit-tested out of the box... it's a huge modular framework to create distributed servers, moreover, it comes with a very complex and large database. However, more and more private communities rely on the fact that open-source projects are not properly fixed, so you can't compete with them. So you need a "leek" or someone like the old "Sunwell" project who suddenly decided to release their hard job.
But that's not what normally happens on open-source projects...
For instance: You don't rely on closed source Oracle DB because MySQL DB is a bugged open-source project. You rely on Oracle DB because it offers some custom/premium features and good customer support. That is exactly what every private project should offer instead and what makes you better than others.

To do it we need more contributors! So, please, join our guild: https://github.com/azerothcore/forum/issues/102 ! We need to expand our team

**Comment by FrancescoBorzi** [comment]
@Yehonal 

I guess the correct title should be "How to complete AzerothCore" without 3.0.

We need to release 3.0 and squash all the pending DB updates because the last time we did it was already more than 1 year ago.

All the above ideas are of course nice, but I don't think we can complete it in a reasonable amount of time.

**Comment by Yehonal** [comment]
> @Yehonal
> 
> I guess the correct title should be "How to complete AzerothCore" without 3.0.
> 
> We need to release 3.0 and squash all the pending DB updates because the last time we did it was already more than 1 year ago.
> 
> All the above ideas are of course nice, but I don't think we can complete it in a reasonable amount of time.

3.0 means: major release -> something that breaks old changes. I would like to stop versioning things in wrong way and keeping trace of PRs/changes that are fixing (patch), adding features (minor), rewriting without legacy support (major)

In this case should a database squash considered a major version? what does it break? If you think so we can decide to proceed in that way otherwise we should use something like 2.5 instead
